src/shared/smbios11.c

```diff
@@ -0,0 +1,61 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "fileio.h"
+#include "macro.h"
+#include "smbios11.h"
+#include "virt.h"
+
+int read_smbios11_field(unsigned i, size_t max_size, char **ret_data, size_t *ret_size) {
+        _cleanup_free_ char *p = NULL, *contents = NULL;
+        _cleanup_free_ void *data = NULL;
+        size_t size, contents_size;
+        int r;
+
+        assert(ret_data);
+        assert(ret_size);
+
+        /* Parses DMI OEM strings fields (SMBIOS type 11), as settable with qemu's -smbios type=11,value=… switch. */
+
+        if (detect_container() > 0) /* don't access /sys/ in a container */
+                return -ENOENT;
+
+        if (asprintf(&p, "/sys/firmware/dmi/entries/11-%u/raw", i) < 0)
+                return -ENOMEM;
+
+        struct dmi_field_header {
+                uint8_t type;
+                uint8_t length;
+                uint16_t handle;
+                uint8_t count;
+                char contents[];
+        } _packed_ *dmi_field_header;
+
+        assert_cc(offsetof(struct dmi_field_header, contents) == 5);
+
+        r = read_virtual_file(
+                        p,
+                        max_size >= SIZE_MAX - offsetof(struct dmi_field_header, contents) ? SIZE_MAX :
+                        sizeof(dmi_field_header) + max_size,
+                        (char**) &data, &size);
```

Comment:
Hmm, why not directly read into `dmi_field_header`?

Reply:
Because the C standard assumes that if you have a pointer to a structure the memory for that structure is fully allocated. But if here, if we'd read the buffer directly into the dmi_field_header pointer we might end up in this scenario that the read is actually shorter than the structure, and that'd be invalid by C standard.

Hence we read into a void* buffer first, then validate the size, and only the convert it to the struct type, since at that point we do know the memory is fully allocated and valid.

Reply:
Yeah, I agree and prefer the current form.